Setting System Feature

afshin opened the pull request:
This PR introduces a setting system for JupyterLab. There are a few core concepts:
* Settings are stored as JSON blobs tied/namespaced to plugins.
* Each plugin's settings have two tiers: `user` and `system`. The `system` JSON represents the default values and is **read-only**. The `user` values are editable by the end user by opening the settings editor from the `Help` menu.

### Change list
- [x] Add the `IDatastore` interface to describe a low-level read/write database. `IStateDB` and `ISettingRegistry` both depend on it. This simplifies swapping out back-ends for state and setting databases, and others if they arise.
- [x] Make `InstanceTracker` independent of `@jupyterlab/application`.
- [x] Create a setting registry extension: `ISettingRegistry`, defined in `@jupyterlab/coreutils`.
- [x] Create a setting system editor.

blink1073 commented on packages/fileeditor-extension/src/index.ts:
```diff
@@ -107,25 +106,18 @@ function activate(app: JupyterLab, registry: IDocumentRegistry, restorer: ILayou
-    if (typeof settings['lineNumbers'] === 'string') {
-      commands.execute(CommandIDs.lineNumbers, settings);
+
+    lineNumbers = !!settings.get('lineNumbers');
```
![noice](https://cloud.githubusercontent.com/assets/2096628/26220490/6a304730-3bd8-11e7-9367-bb950f11fc0a.png)

ellisonbg commented:
Let us know when this is ready for review ;-)

afshin commented:
@ellisonbg I will let you know. And I'll write it up so that it's not just a big blob of code.

blink1073 commented on packages/apputils-extension/src/index.ts:
```diff
@@ -104,6 +104,21 @@ const palettePlugin: JupyterLabPlugin<ICommandPalette> = {
+  id: 'jupyter.services.setting-registry',
+  requires: [IStateDB],
+  activate: (app: JupyterLab, state: IStateDB) => {
+    const registry = new SettingRegistry();
```
Use the statedb here?

afshin replied:
This happens automatically inside the `SettingRegistry` constructor:

```typescript
/**
 * Create a new setting registry.
 */
constructor(options: SettingRegistry.IOptions = { datastore: null }) {
  const namespace = 'jupyter.db.settings';
  this._datastore = options.datastore || new StateDB({ namespace });
}
```

blink1073 replied:
What I mean is that the state instance should be used (`new SettingRegistry({ datastore: state })`)

blink1073 replied:
Either that or the plugin shouldn't depend on a statedb.

afshin replied:
Ah right, good point. I originally did it that way and then I forgot to remove `IStateDB` as a requirement when I switched it.

afshin commented:
FYI @ellisonbg, I had some merge conflicts when I rebased, so while I was resolving them, I also refactored this file which may affect work you want to do: https://github.com/jupyterlab/jupyterlab/pull/2157/commits/e7dbedce921d5a329ce1f9ab3fa8eaab55add8c5

ellisonbg commented:
@vineetgorhe would also affect you if you dive into the markdown stuff again.

blink1073 commented on packages/apputils-extension/src/index.ts:
```diff
@@ -101,6 +104,17 @@ const palettePlugin: JupyterLabPlugin<ICommandPalette> = {
 
 
 /**
+ * The default layout restorer provider.
```
incorrect description

mpacer commented:
Just a note about an example of a SublimeText Package that has an extremely complicated settings format with tonnes of documentation: https://github.com/SublimeText/LaTeXTools.

It's not the case that we should design for, but it's something that is likely to come up if people are doing complicated publishing oriented tasks like writing papers/reports.

blink1073 commented:
Looks and works great!
